### PR TITLE
forkchoice.Getter wrapper with locking wrappers

### PR DESCRIPTION
**Pull request by kasey**
This PR adds a new type `ROForkchoice` which implements `forkchoice.Getter`, calling `Rlock`/`RUnlock` around each call delegated to the wrapped forkchoice implementation. This is intended to solve the problem of ensuring that code outside of the blockchain is only able to access forkchoice in a threadsafe fashion.

### CODE REVIEW

**Review by potuz — COMMENTED** [review]
I don't really grasp the point of this PR. As I had understood it before, you had mentioned that we'd implement a getter that would return a Getter interphase for forkchoice that will come back locked. Callers would obtain a self-locked forkchoice to call whatever method the Getter interface exposes.  Part of the benefits of this is that the callers would not need to deal with the wrappers in `chain-info.go`. 

However in this PR implements the same wrappers that are there in chain-info, explicitly locking forkchoice and calling the corresponding wrapped methods. In practical terms, for the caller, why is this better than simply calling chain-info?

File: beacon-chain/forkchoice/ro.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	fieldparams "github.com/prysmaticlabs/prysm/v4/config/fieldparams"
+	forkchoice2 "github.com/prysmaticlabs/prysm/v4/consensus-types/forkchoice"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+)
+
+// ROForkChoice is an implementation of forkchoice.Getter which calls `Rlock`/`RUnlock`
+// around a delegated method call to the underlying Getter implementation.
+type ROForkChoice struct {
+	getter Getter
+	l      Locker
```
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/forkchoice/ro.go:16]: I could split out locker into a narrower interface and move the `Lock`/`Unlock` methods into the main `ForkChoicer`, I just thought putting all the lock related stuff into a single interface was more tidy.

File: beacon-chain/forkchoice/ro.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+var _ Getter = &ROForkChoice{}
+
+// ROWrappable represents the subset of ForkChoicer a type needs to support
+// in order for ROForkChoice to wrap it. This simplifies the creation of a mock
+// type that can be used to assert that all of the wrapped methods are correctly
+// called between mutex acquire/release.
+type ROWrappable interface {
+	Locker
```
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/forkchoice/ro.go]: updated

File: beacon-chain/forkchoice/ro.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/forkchoice/ro.go]: updated

File: beacon-chain/forkchoice/interfaces.go [hunk 3]
```diff
@@ -47,29 +52,33 @@ type AttestationProcessor interface {
...
-	ProposerBoost() [fieldparams.RootLength]byte
+	FastGetter
 	AncestorRoot(ctx context.Context, root [32]byte, slot primitives.Slot) ([32]byte, error)
 	CommonAncestor(ctx context.Context, root1 [32]byte, root2 [32]byte) ([32]byte, primitives.Slot, error)
-	IsCanonical(root [32]byte) bool
+	ForkChoiceDump(context.Context) (*forkchoice2.Dump, error)
+	Tips() ([][32]byte, []primitives.Slot)
+}
+
+type FastGetter interface {
```
Comment (potuz) [inline beacon-chain/forkchoice/interfaces.go:62]: Thanks for ordering these

**Review by potuz — DISMISSED** [review]
LGTM, I think when we fix the `rpc` package to use these we will need to expose a slow lock on `Head()`

File: beacon-chain/forkchoice/ro.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+// LastRoot delegates to the underlying forkchoice call, under a lock.
+func (ro *ROForkChoice) LastRoot(e primitives.Epoch) [32]byte {
+	ro.l.RLock()
+	defer ro.l.RUnlock()
+	return ro.getter.LastRoot(e)
+}
+
+func (ro *ROForkChoice) TargetRoot(root [32]byte) ([32]byte, error) {
```
Comment (terencechain) [inline beacon-chain/forkchoice/ro.go:159]: Missing a godoc comment here

Comment (kasey) [reply beacon-chain/forkchoice/ro.go:160]: updated

**Review by terencechain — DISMISSED** [review]

**Review by terencechain — APPROVED** [review]

**Pull request merged**